test/e2e/framework/service/jig.go

```diff
@@ -1051,3 +958,121 @@ func (j *TestJig) CreateSCTPServiceWithPort(tweak func(svc *v1.Service), port in
+			framework.Logf("skipping testEndpointReachability() for internal adddress %s", internalAddr)
+			continue
+		}
+		isNodeInternalIPV4 := net.ParseIP(internalAddr).To4() != nil
+		// Check service reachability on the node internalIP which is same family
+		// as clusterIP
+		if isClusterIPV4 != isNodeInternalIPV4 {
+			framework.Logf("skipping testEndpointReachability() for internal adddress %s as it does not match clusterIP (%s) family", internalAddr, clusterIP)
+			continue
+		}
+
+		err := j.testEndpointReachability(internalAddr, sp.NodePort, sp.Protocol, pod)
+		if err != nil {
+			return err
+		}
+	}
+	for _, externalAddr := range externalAddrs {
+		isNodeExternalIPV4 := net.ParseIP(externalAddr).To4() != nil
+		if isClusterIPV4 != isNodeExternalIPV4 {
+			framework.Logf("skipping testEndpointReachability() for external adddress %s as it does not match clusterIP (%s) family", externalAddr, clusterIP)
+			continue
+		}
+		err := j.testEndpointReachability(externalAddr, sp.NodePort, sp.Protocol, pod)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// isInvalidOrLocalhostAddress returns `true` if the provided `ip` is either not
+// parsable or the loopback address. Otherwise it will return `false`.
+func isInvalidOrLocalhostAddress(ip string) bool {
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil || parsedIP.IsLoopback() {
+		return true
+	}
+	return false
+}
+
+// testEndpointReachability tests reachability to endpoints (i.e. IP, ServiceName) and ports.  Note that the execPod
+// must have nc installed.
+func (j *TestJig) testEndpointReachability(endpoint string, port int32, protocol v1.Protocol, execPod *v1.Pod) error {
+	// 1) Wait for the pod to come up before we test endpoints
+	e2epod.WaitForPodCondition(j.Client, execPod.ObjectMeta.Namespace, execPod.ObjectMeta.Name, "", ServiceReachabilityShortPollTimeout,
+		func(pod *v1.Pod) (bool, error) {
+			framework.Logf("Waiting for pod %v to be running, current state %v", execPod.ObjectMeta.Name, pod.Status.Phase)
+			return pod.Status.Phase == v1.PodRunning, nil
+		})
+
+	// 2) make sure the exec pod is valid before attempting to poll nc, otherwise when misusing this function
+	// alot of spam can get created.
+	err := wait.PollImmediate(1*time.Second, ServiceReachabilityShortPollTimeout, func() (bool, error) {
+		if _, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, "which nc"); err != nil {
+			framework.Logf("verifying execPod can run 'nc': %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("nc must be installed on the target pod for testEndpointReachability to work properly")
+	}
+
+	// 3) Finally take care of the polling
+	ep := net.JoinHostPort(endpoint, strconv.Itoa(int(port)))
+	cmd := ""
+	switch protocol {
+	case v1.ProtocolTCP:
+		cmd = fmt.Sprintf("nc -zv -t -w 2 %s %v", endpoint, port)
+	case v1.ProtocolUDP:
+		cmd = fmt.Sprintf("nc -zv -u -w 2 %s %v", endpoint, port)
```

Comment:
Why are we using the -z option here? Is there a noted reason to not send data? Also I'm wondering if we want to move away from netcat in general in favor of agnhost. Agnhost seems to better align with the way many connections in NetPol are tested [test/e2e/network/network_policy.go](https://github.com/kubernetes/kubernetes/blob/1c403c311e80dcdb075219d117ac803f8891de39/test/e2e/network/network_policy.go#L2049-L2051)

Reply:
In favor of agn host but I thought we still used nc in various places.  As long as we do use it.  The ability to use NCat (as opposed to the old netcat) gives us parity between sctp udp tcp ....

Reply:
A quck scan for `nc -` matches in the codebase point toward main uses for udp port checking. This likely is due to agnhost connect cannot be used for UDP.